proxy/coco/src/request/states.rs

```diff
@@ -0,0 +1,308 @@
+    }
+}
+
+/// The number of different attempts a `Request` has made during its lifetime.
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Attempts {
+    /// The number of query attempts we have made.
+    pub(super) queries: Queries,
+    /// The number of clone attempts we have made.
+    pub(super) clones: Clones,
+}
+
+impl Attempts {
+    /// Get a new `Attempts` where the number of queires and clones is initially `0`.
+    #[must_use]
+    pub const fn new() -> Self {
+        Attempts {
+            queries: Queries(0),
+            clones: Clones(0),
+        }
+    }
+}
+
+impl Default for Attempts {
+    fn default() -> Self {
+        Attempts::new()
+    }
+}
+
+impl sealed::Sealed for Attempts {}
+
+// State Traits
+
+/// If a state type implements this trait then it means that the type is allowed to increment the
+/// query attempts in a `Request`.
+///
+/// The trait is sealed internally, so we do not expect end-users to implement it.
+pub trait QueryAttempt: sealed::Sealed {}
+impl QueryAttempt for IsRequested {}
+
+/// If a state type implements this trait then it means that the type holds a `HashMap` of peers and
+/// their status of cloning.
+///
+/// The trait is sealed internally, so we do not expect end-users to implement it.
+pub trait HasPeers: sealed::Sealed
+where
+    Self: Sized,
+{
+    /// Give back the underlying `HashMap` of peers that is contained in `Self`.
+    fn peers(&mut self) -> &mut HashMap<PeerId, Status>;
+
+    /// Returns `false` if the `peers` are empty or if any of them are `Status::Available`
+    /// or `Status::InProgress`.
+    ///
+    /// Otherwise, if all are in the `Status::Failed` state, then we return `true`.
+    fn all_failed(&mut self) -> bool {
+        let peers = self.peers();
+
+        if peers.is_empty() {
+            return false;
+        }
+
+        peers.iter().all(|(_, status)| *status == Status::Failed)
+    }
+}
+
+impl HasPeers for Found {
+    fn peers(&mut self) -> &mut HashMap<PeerId, Status> {
+        &mut self.peers
+    }
+}
+
+impl HasPeers for Cloning {
+    fn peers(&mut self) -> &mut HashMap<PeerId, Status> {
+        &mut self.peers
+    }
+}
+
+/// If a state type implements this trait it means that their is a valid transition from that state
+/// to the `IsCanceled` state.
+///
+/// The trait is sealed internally, so we do not expect end-users to implement it.
+pub trait Cancel: sealed::Sealed
+where
+    Self: Sized,
+{
+    /// Transition the state into `IsCanceled`. This ignores whatever state we were in and defaults
+    /// by returning the `IsCanceled` state.
+    fn cancel(self) -> IsCanceled {
+        PhantomData
+    }
+}
+
+impl Cancel for IsCreated {}
+impl Cancel for IsRequested {}
+impl Cancel for Found {}
+impl Cancel for Cloning {}
+impl Cancel for IsCanceled {}
+
+/// If a state type implements this trait it means that their is a valid transition from that state
```

Comment:
Also here, what did you mean by "their"?

Reply:
copy-pasta